include/wil/safearrays.h

```diff
@@ -0,0 +1,897 @@
+    //!     {
+    //!         result.emplace_back(wil::make_bstr_nothrow(bstr));
+    //!     }
+    //!     return S_OK;
+    //! }
+    //! ~~~~
+    //! unique_bstr_safearray move_to_safearray(std::vector<wil::unique_bstr>& source)
+    //! {
+    //!     auto sa = wil::unique_bstr_safearray{source.size()};
+    //!     auto current = std::begin(source);
+    //!     for(BSTR& bstr : sa.access_data())
+    //!     {
+    //!         // Transfer ownership of the BSTR into the safearray
+    //!         bstr = (*current++).release();
+    //!     }
+    //!     return sa;
+    //! }
+    //! ~~~~
+    //! void handle_safearray_of_bools(SAFEARRAY* psa, std::function<void(bool, ULONG)> fnHandler)
+    //! {
+    //!     auto data = wil::safearraydata<VARIANT_BOOL>{psa};
+    //!     for (auto i = 0U; i < data.size(); ++i)
+    //!     {
+    //!         fnHandler(data[i] != VARIANT_FALSE, i);
+    //!     }
+    //! }
+    //! ~~~~
+    //! template<typename T>
+    //! std::vector<T> extract_all_values(SAFEARRAY* psa)
+    //! {
+    //!     auto data = wil::safearraydata<T>{psa};
+    //!     // Construct the vector from the iterators
+    //!     return {data.begin(), data.end()};
+    //! }
+    //! ~~~~
+    //! template<typename INTERFACE>
+    //! std::vector<wil::com_ptr<T>> extract_all_interfaces(SAFEARRAY* psa)
+    //! {
+    //!     auto result = std::vector<wil::com_ptr<INTERFACE>>{};
+    //!     auto data = wil::safearraydata<LPUNKNOWN>{psa};
+    //!     result.reserve(data.size());
+    //!     for(auto& p : data) // type of P is LPUNKNOWN
+    //!     {
+    //!         // Use "tag_com_query" if you want failure instead of nullptr if INTERFACE not supported
+    //!         result.emplace_back(p, details::tag_try_com_query);
+    //!     }
+    //!     return result;
+    //! }
+    //! ~~~~
+    //! @tparam T           The underlying datatype expected in the SAFEARRAY. It should be a scalar type, BSTR, VARIANT_BOOL, LPUNKNOWN/LPDISPATCH, etc.
+    //! @tparam err_policy  Represents the error policy for the class (error codes, exceptions, or fail fast; see @ref page_errors)
+    template <typename T, typename err_policy = err_exception_policy>
+    class safearraydata_t
+    {
+    private:
+        typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnaccessData), details::SafeArrayUnaccessData, details::pointer_access_noaddress> safearray_unaccess_data;
+
+    public:
+        typedef typename err_policy::result result;
+        typedef typename safearray_unaccess_data::pointer pointer;
+        typedef T value_type;
+        typedef value_type* value_pointer;
+        typedef const value_type* const_value_pointer;
+        typedef value_type& value_reference;
+        typedef const value_type& const_value_reference;
+        typedef value_type* iterator;
+        typedef const value_type* const_iterator;
+
+    public:
+        // Exception-based construction
+        safearraydata_t(pointer psa)

```

Comment:
Probably want this to be `explicit`

Reply:
Yes!  This absolutely should be explicit!  Will fix.

Reply:
There was also a constructor in `safearray_t` that took 2 parameters (but one has a default value) that also needed to be explicit as well.